pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1156 @@
+// Caller must always pass in a list of existing nodes to the new allocator.
+// NodeList is only nil in testing.
+func NewMultiCIDRRangeAllocator(
+		client clientset.Interface,
+		nodeInformer informers.NodeInformer,
+		clusterCIDRInformer networkinginformers.ClusterCIDRInformer,
+		allocatorParams CIDRAllocatorParams,
+		nodeList *v1.NodeList,
+		cccList *networkingv1alpha1.ClusterCIDRList,
+		testCIDRMap map[string][]*cidrset.ClusterCIDR,
+) (CIDRAllocator, error) {
+	if client == nil {
+		klog.Fatalf("client is nil")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventSource := v1.EventSource{
+		Component: "multiCIDRRangeAllocator",
+	}
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, eventSource)
+	eventBroadcaster.StartStructuredLogging(0)
+	klog.V(0).Infof("Started sending events to API Server. (EventSource = %v)", eventSource)
+
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{
+			Interface: client.CoreV1().Events(""),
+		})
+
+	if client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		ratelimiter.RegisterMetricAndTrackRateLimiterUsage("multi_cidr_range_allocator", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	ra := &multiCIDRRangeAllocator{
+		client:                client,
+		nodeLister:            nodeInformer.Lister(),
+		nodesSynced:           nodeInformer.Informer().HasSynced,
+		clusterCIDRLister:     clusterCIDRInformer.Lister(),
+		clusterCIDRSynced:     clusterCIDRInformer.Informer().HasSynced,
+		nodeCIDRUpdateChannel: make(chan multiCIDRNodeReservedCIDRs, cidrUpdateQueueSize),
+		recorder:              recorder,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "multi_cidr_range_allocator"),
+		lock:                  &sync.Mutex{},
+		nodesInProcessing:     sets.NewString(),
+		cidrMap:               make(map[string][]*cidrset.ClusterCIDR, 0),
+	}
+
+	// testCIDRMap is only set for testing purposes.
+	// Warning: If testCIDRMap is set in production, it might be a misconfiguration
+	// or a bug.
+	if len(testCIDRMap) > 0 {
+		ra.cidrMap = testCIDRMap
+	}
+
+	if cccList == nil {
+		cccList = &networkingv1alpha1.ClusterCIDRList{}
+	}
+	createDefaultClusterCIDR(cccList, allocatorParams)
+
+	// Regenerate the cidrMaps from the existing ClusterCIDRs.
+	for _, clusterCIDR := range cccList.Items {
+		klog.Infof("Regenerating existing ClusterCIDR: %v", clusterCIDR)
+		// Create an event for invalid ClusterCIDRs, do not crash on failures.
+		if err := ra.reconcileBootstrap(&clusterCIDR); err != nil {
+			klog.Errorf("Error while regenerating existing ClusterCIDR: %v", err)
+			ra.recorder.Event(&clusterCIDR, "Warning", "InvalidClusterCIDR", err.Error())
+		}
+	}
+
+	clusterCIDRInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    createClusterCIDRHandler(ra.reconcileCreate),
+		DeleteFunc: createClusterCIDRHandler(ra.reconcileDelete),
```

Comment:
Not clear why we call reconcileDelete on the object deletion? Isn't the object already gone at this point/